source/Octopus.Tentacle.Tests.Integration/Util/Builders/Decorators/Proxies/MethodTracingStats.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Threading;
+
+namespace Octopus.Tentacle.Tests.Integration.Util.Builders.Decorators.Proxies
+{
+    public class MethodTracingStats : IRecordedMethodTracingStats
+    {
+        readonly ConcurrentDictionary<string, Lazy<TracedMethodStats>> trackedMethods = new();
+
+        IRecordedTracedMethodStats IRecordedMethodTracingStats.For(string methodName) => GetMethodStats(methodName);
+
+        public void RecordCallStart(MethodInfo targetMethod)
+        {
+            var stats = GetMethodStats(targetMethod);
+            stats.RecordStarted();
+        }
+
+        public void RecordCallComplete(MethodInfo targetMethod)
+        {
+            var stats = GetMethodStats(targetMethod);
+            stats.RecordCompleted();
+        }
+
+        public  void RecordCallException(MethodInfo targetMethod, Exception exception)
+        {
+            var stats = GetMethodStats(targetMethod);
+            stats.RecordException(exception);
+        }
+
+        TracedMethodStats GetMethodStats(string methodName) => trackedMethods.GetOrAdd(NormalizeMethodName(methodName), _ => new Lazy<TracedMethodStats>(() => new TracedMethodStats())).Value;
+        TracedMethodStats GetMethodStats(MethodInfo targetMethod) => GetMethodStats(targetMethod.Name);
+
+        static string NormalizeMethodName(string methodName)
```

Comment:
Normalising just for sync/async service method names (as the extension method uses the same object for both). Can be removed when sync services are removed